src/main/scala/riscv/core/fivestage/CPU.scala

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 Howard Lau
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package riscv.core.fivestage
+
+import bus.{AXI4LiteChannels, AXI4LiteMaster}
+import chisel3._
+import riscv.Parameters
+
+class CPU extends Module {
+  val io = IO(new Bundle {
+    val axi4_channels = new AXI4LiteChannels(Parameters.AddrBits, Parameters.DataBits)
+    val bus_address = Output(UInt(Parameters.AddrWidth))
+    val interrupt_flag = Input(UInt(Parameters.InterruptFlagWidth))
+    val stall_flag_bus = Input(Bool())
+    val debug_read_address = Input(UInt(Parameters.PhysicalRegisterAddrWidth))
+    val debug_read_data = Output(UInt(Parameters.DataWidth))
+
+    val instruction_valid = Input(Bool())
+  })
+
+  val ctrl = Module(new Control)
+  val regs = Module(new RegisterFile)
+  val inst_fetch = Module(new InstructionFetch)
+  val if2id = Module(new IF2ID)
+  val id = Module(new InstructionDecode)
+  val id2ex = Module(new ID2EX)
+  val ex = Module(new Execute)
+  val ex2mem = Module(new EX2MEM)
+  val mem = Module(new MemoryAccess)
+  val mem2wb = Module(new MEM2WB)
+  val wb = Module(new WriteBack)
+  val forwarding = Module(new Forwarding)
+  val clint = Module(new CLINT)
+  val csr_regs = Module(new CSR)
+  val axi4_master = Module(new AXI4LiteMaster(Parameters.AddrBits, Parameters.DataBits))
+
+  axi4_master.io.channels <> io.axi4_channels
+
+  // The EX module takes precedence over IF (but let the previous fetch finish)
+  val ex_granted = RegInit(false.B)
+  when(ex_granted) {
```

Comment:
It may be better to rename it to `mem_granted` since we are not doing memory operations in EX now.